### PR TITLE
ENH: add y-intercept for logistic regression

### PR DESCRIPTION
**Pull request by facaiy**
Hi, the PR is aimed at adding intercept for logistic regression #142 .

The work is not done yet (mostly finished in fact), but feedback is necessary for me to determine whether to go further or not. Hence, the PR is opened.

### Mathematical Derivation

Given:
+  linear model: f(x) = w^T x + b
+  loss function: L(y, f(x))

the derivative of b is: \frac{dL}{df}

### Implementation

The PR is inspired by:
+ spark: [LogisticRegression.scala](https://github.com/apache/spark/blob/cb8d5cc90ff8d3c991ff33da41b136ab7634f71b/mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala#L1699)
+ sklearn: [logistic.py](https://github.com/scikit-learn/scikit-learn/blob/924a4661ee3e342686b89eb47c12a6e660585376/sklearn/linear_model/logistic.py#L126)

### How to test

+ [x] code review by expert.
+ [x] Perhaps some tests are needed, however, more feedback is needed for me now.

### CODE REVIEW

**Comment by TAAAN** [comment]
Thanks for your PR. 
When preprocessing train data, set the fist dim 1 for every instance. Then the fist dim of w is b.

**Comment by andyyehoo** [comment]
@facaiy it's a good idea. Just go on with this PR and finish it. We will review and merge it with certain modification.

**Comment by facaiy** [comment]
OK. I'll ping you when finished.

**Comment by facaiy** [comment]
The PR is tested on [a4a](https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/) dataset.

There are two questions:

1. The package failed to submit its tasks to YARN. I believe the bug is introduced from the latest branch-1.0.0, instead of the PR. I will create a new issue to report it later.

2. I picked those commits to `commit 3ff4329825fd355118384d5be438dd132f1015c3`, and everything is OK. However, the result of experiment on a4a dataset implies that AUC decreased when y-intercept is used.

use y-intercept | iteration | auc | duration | reg_l2 | learn rate
---|----|----------------------|------|----|-----
No | 10 | 0.8991570991465441 | 47s | 0 | 4
Yes | 10 | 0.898142663326485 | 50s | 0 | 4

**Comment by facaiy** [comment]
Many thanks, @andyyehoo @TAAAN .
